### ISSUE TITLE
v2.23 frequently allocates and frees device/managed memory

### ISSUE THREAD

**Issue by hengjiew** [hypre-space/hypre#507]
Hello, I am trying to use BoomerAMG preconditioned GMRES in hypre on the Summit cluster. I was able to use the hypre/2.20.0 module provided by Summit. When I installed the latest hypre (commit aadfd86d) and built with the following options,
```
./configure --with-cuda --enable-unified-memory --enable-device-memory-pool
make install
```
the performance drops drastically compared to using hypre/2.20. When I used nsight-systems to profile the code, it showed that there were a large number of `cudaMalloc` and `cudaFree` called in v2.23 but almost none called in v2.20. 
May I ask if I am missing anything in configuring hypre? Thanks!

**Comment by liruipeng** [comment]
@hengjiew  Can you please clarify if PFMG is being used or BoomerAMG, and with which you saw the large number of malloc/free?  Thanks!

**Comment by hengjiew** [comment]
> @hengjiew Can you please clarify if PFMG is being used or BoomerAMG, and with which you saw the large number of malloc/free? Thanks!

Sorry for the confusion. I correct it in the original question: I was using BoomerAMG preconditioned GMRES. (PFMG was not used.) More specifically, I tested hypre inside AMReX, an AMR package that supports using hypre to solve linear systems. I think the most time-consuming function here is `HYPRE_ParCSRGMRESSolve`, which incurs lots of the `malloc/free`.

**Comment by liruipeng** [comment]
@hengjiew Thanks for the info. What was the CUDA version used?

**Comment by hengjiew** [comment]
> @hengjiew Thanks for the info. What was the CUDA version used?

I am using cuda 11.4.0.

**Comment by liruipeng** [comment]
> > @hengjiew Thanks for the info. What was the CUDA version used?
> 
> I am using cuda 11.4.0.

OK. The reason of malloc/free is probably from the matvec with cuSPARSE > 11 where we allocate buffer and free at each time. (The CUDA 11 support was added after hypre 2.20). We should optimize this. But I am not sure why hypre's memory pool did not handle the malloc properly. At the meantime, maybe you want to try using [Umpire](https://github.com/LLNL/Umpire) which in general gives much better performance for hypre on GPUs. Thanks!

**Comment by hengjiew** [comment]
Thanks, @liruipeng! 
I observed the same performance drop with cuda 11.0.3 (first observed with cuda 11.4.0). Then I build hypre with umpire by specifying 
```
--with-umpire
--with-umpire-include=myinclude
--with-umpire-lib-dirs=mylib
--with-umpire-libs=umpire
```
However, the performance didn't get improved. May I ask if I am using the right options here?

**Comment by WeiqunZhang** [comment]
@liruipeng I mentioned this to @ulrikeyang a while ago, and she suggested that I should talk to you.  But I never got around to it.  Is it possible for hypre to use an external memory allocator provided by the user at runtime?  In AMReX, we already have a memory pool.  It would be great if hypre can allow an AMReX app to provide an allocator.  We do this with SUNDIALS.  See https://github.com/AMReX-Codes/amrex/blob/development/Src/Extern/SUNDIALS/AMReX_SUNMemory.cpp.

**Comment by PaulMullowney** [comment]
Don't you have to configure the size of the Umpire memory pool in order to ensure that it's properly handled. If the problem is very large, then the pool might not be big enough?

**Comment by WeiqunZhang** [comment]
A question, does hypre provide version macros?  There is only one place in AMReX where 2.23 is needed for ILU smoothing.  We could #if that out for people who do not need that feature.

**Comment by liruipeng** [comment]
> @liruipeng I mentioned this to @ulrikeyang a while ago, and she suggested that I should talk to you. But I never got around to it. Is it possible for hypre to use an external memory allocator provided by the user at runtime? In AMReX, we already have a memory pool. It would be great if hypre can allow an AMReX app to provide an allocator. We do this with SUNDIALS. See https://github.com/AMReX-Codes/amrex/blob/development/Src/Extern/SUNDIALS/AMReX_SUNMemory.cpp.

You may try this 
https://github.com/hypre-space/hypre/blob/5262bff46124777f81fbdff5bf967ea20a037df2/src/utilities/handle.c#L120
I don't think anyone has tried yet. Only for GPU **device-only** memory. Let me know if it can work for you.

Update: link corrected

**Comment by WeiqunZhang** [comment]
Great! We will try.  Also I see there is `#define HYPRE_RELEASE_NUMBER` in HYPRE_config.h.  We could also use that to allow @hengjiew to use an older version of hypre.

**Comment by liruipeng** [comment]
> A question, does hypre provide version macros? There is only one place in AMReX where 2.23 is needed for ILU smoothing. We could #if that out for people who do not need that feature.

Yes, check `#define HYPRE_RELEASE_NUMBER 22201`

**Comment by liruipeng** [comment]
> Don't you have to configure the size of the Umpire memory pool in order to ensure that it's properly handled. If the problem is very large, then the pool might not be big enough?

Good point. It could be. Here are the defaults of Umpire settings

https://github.com/hypre-space/hypre/blob/5262bff46124777f81fbdff5bf967ea20a037df2/src/utilities/general.c#L318

**Comment by liruipeng** [comment]
> Thanks, @liruipeng! I observed the same performance drop with cuda 11.0.3 (first observed with cuda 11.4.0). Then I build hypre with umpire by specifying
> 
> ```
> --with-umpire
> --with-umpire-include=myinclude
> --with-umpire-lib-dirs=mylib
> --with-umpire-libs=umpire
> ```
> 
> However, the performance didn't get improved. May I ask if I am using the right options here?

If possible and you don't mind, can you try CUDA 10 to confirm the matvec is really the issue? With CUDA 10, you should not see those many malloc/free, if that's the culprit . Thanks again!

**Comment by liruipeng** [comment]
> Thanks, @liruipeng! I observed the same performance drop with cuda 11.0.3 (first observed with cuda 11.4.0). Then I build hypre with umpire by specifying
> 
> ```
> --with-umpire
> --with-umpire-include=myinclude
> --with-umpire-lib-dirs=mylib
> --with-umpire-libs=umpire
> ```
> 
> However, the performance didn't get improved. May I ask if I am using the right options here?

And we will push a fix for it soon.

**Comment by liruipeng** [comment]
@hengjiew Hello again. Can you please check out this branch https://github.com/hypre-space/hypre/tree/cusparseNewSpMV and test if the problem is still there?

Thanks!

**Comment by hengjiew** [comment]
> @hengjiew Hello again. Can you please check out this branch https://github.com/hypre-space/hypre/tree/cusparseNewSpMV and test if the problem is still there?
> 
> Thanks!

Hello, I build the new branch with the following options:
```
./configure --with-cuda --enable-unified-memory --enable-device-memory-pool --with-cuda-home=$CUDAPATH
```
Then I got the following error during compilation:
```
/sw/summit/cuda/11.0.3/bin/nvcc -ccbin=mpixlC -gencode arch=compute_70,code=sm_70  -O2 -lineinfo -expt-extended-lambda -std=c++11 --x cu -Xcompiler "-O2 " -DHAVE_CONFIG_H -I.. -I. -I./.. -I./../utilities  -I/sw/summit/cuda/11.0.3/include       -c csr_matvec_device.c -o csr_matvec_device.obj
csr_matvec_device.c(164): error: identifier "CUSPARSE_SPMV_CSR_ALG2" is undefined
csr_matvec_device.c(180): error: identifier "CUSPARSE_SPMV_CSR_ALG2" is undefined
2 errors detected in the compilation of "csr_matvec_device.
```
My environment:  cuda/11.0.3, gcc/9.3.0, and spectrum-mpi/10.4.0.3-20210112.

**Comment by liruipeng** [comment]
> > @hengjiew Hello again. Can you please check out this branch https://github.com/hypre-space/hypre/tree/cusparseNewSpMV and test if the problem is still there?
> > Thanks!
> 
> Hello, I build the new branch with the following options:
> 
> ```
> ./configure --with-cuda --enable-unified-memory --enable-device-memory-pool --with-cuda-home=$CUDAPATH
> ```
> 
> Then I got the following error during compilation:
> 
> ```
> /sw/summit/cuda/11.0.3/bin/nvcc -ccbin=mpixlC -gencode arch=compute_70,code=sm_70  -O2 -lineinfo -expt-extended-lambda -std=c++11 --x cu -Xcompiler "-O2 " -DHAVE_CONFIG_H -I.. -I. -I./.. -I./../utilities  -I/sw/summit/cuda/11.0.3/include       -c csr_matvec_device.c -o csr_matvec_device.obj
> csr_matvec_device.c(164): error: identifier "CUSPARSE_SPMV_CSR_ALG2" is undefined
> csr_matvec_device.c(180): error: identifier "CUSPARSE_SPMV_CSR_ALG2" is undefined
> 2 errors detected in the compilation of "csr_matvec_device.
> ```
> 
> My environment: cuda/11.0.3, gcc/9.3.0, and spectrum-mpi/10.4.0.3-20210112.

Fixed. Sorry about that.

**Comment by hengjiew** [comment]
Thanks @liruipeng 
Below is the time cost for solving a Poisson equation on 6 GPUs. The time is the total of some MG levels (same for each test) + using hypre as the bottom solver. 
``` 
version     time(s)
v2.23       0.81
your fix    0.39
v2.20       0.28
```
Your fix definitely improves the performance but it is still slower than v2.20 to some extent.

**Comment by liruipeng** [comment]
> Thanks @liruipeng Below is the time cost for solving a Poisson equation on 6 GPUs. The time is the total of some MG levels (same for each test) + using hypre as the bottom solver.
> 
> ```
> version     time(s)
> v2.23       0.81
> your fix    0.39
> v2.20       0.28
> ```
> 
> Your fix definitely improves the performance but it is still slower than v2.20 to some extent.

Did you use CUDA 11 with hypre v2.20?

**Comment by hengjiew** [comment]
I was using the system-provided hypre/2.20 and cuda 11.0.3 on Summit. I assumed the system would load a compatible hypre library with that cuda. However, the following command outputs nothing.
```
ldd libHYPRE.so | grep cu
```
Also  HYPRE_config.h sets  `#define HYPRE_USING_HOST_MEMORY 1` but no `HYPRE_USING_CUDA`. So the reason why I didn't observe any malloc/free with v2.20 is probably that it didn't use GPU. But wouldn't the GPU version (with your fix) be faster than v2.20 on CPU?

**Comment by liruipeng** [comment]
> I was using the system-provided hypre/2.20 and cuda 11.0.3 on Summit. I assumed the system would load a compatible hypre library with that cuda. However, the following command outputs nothing.
> 
> ```
> ldd libHYPRE.so | grep cu
> ```
> 
> Also HYPRE_config.h sets `#define HYPRE_USING_HOST_MEMORY 1` but no `HYPRE_USING_CUDA`. So the reason why I didn't observe any malloc/free with v2.20 is probably that it didn't use GPU. But wouldn't the GPU version (with your fix) be faster than v2.20 on CPU?

Thanks for checking that.

I don't know about the hypre available on Summit, but it seems that it wasn't configured with CUDA. Also, another evidence, v2.20 should not work with CUDA 11 (compile error).

On the speed of hypre on GPU vs. CPU, it depends on problem sizes, AMG parameters chosen, etc. We cannot guarantee that the GPU version is faster than CPU always. 

There are many moving parts here. We need to know more details to figure out what it is actually doing. Thanks.

**Comment by hengjiew** [comment]
That makes sense. I guess we don't need to dwell on the CPU/GPU comparison for this particular test.  
Thanks for the fix!
